File: vms/compile-rust/wes-gpio/ds18b20/ds18b20.py
# sudo apt install -y python3-libgpiod # for system packages, can this work in a venv?
#   seems like I might need apt to install deps for gpiod... version in venv doesn't work but then if I move out of venv it works
# https://pypi.org/project/gpiod/

# !!! TIMING MIGHT BE WORKING in my python!
# !!! ok my timing looks close enough now actually, so I might have smth wrong with the protocol, let's revisit that next ... I am within the tolerances I understand and still don't get a response to the read ROM... I bet I have smth off in my understanding
#  DO NOT BE SLOPPY with prints or anything that adds overhead... one print can add 50us, same with not carefully choosing when to drive line low (before/after requesting output mode)
# PRN might be able to keep a list of messages and dump on a failure/completino as long as adding to the list is trivial <1us timing
# !!! WES KEEP IN MIND timing might not work out in python still... c code would rock and you can have logging there that doesn't kill perf (i.e. pr_ dev_ printk_ in a module)

# FYI I checked out libgpiod for the APIs => https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.gitß => libgpiod/bindings/python

import logging
import math
import gpiod
import time
from gpiod.line import Direction, Value

import crcmod  # ~1ms

ds18b20_crc8 = crcmod.mkCrcFun(0x131, initCrc=0, xorOut=0)  # ~0.250us, so only do it once... this is reusable right?

CHIP = 'gpiochip0'
DS1820B_PIN = 12

# readability:
LOW = Value.INACTIVE
HIGH = Value.ACTIVE

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.ERROR)

# ! NOTE ABOUT gpiod pkg (in venv vs system pkgs):
# ! below uses gpiod via `pip install gpiod` via venv
# !   IS NOT same as gpiod pkg via `apt install apt info python3-libgpiod`
# research on diffs:
# system pkgs => python3 =>
#   import gpiod
#   gpiod.version_string() #  1.6.3
#   help(gpiod)   # completely diff APIs
#   gpiod # <module 'gpiod' from '/usr/lib/python3/dist-packages/gpiod.cpython-311-aarch64-linux-gnu.so'>
#
#   exit repl:
#       dpkg -S /usr/lib/python3/dist-packages/gpiod.cpython-311-aarch64-linux-gnu.so
#       # => python3-libgpiod:arm64: /usr/lib/python3/dist-packages/gpiod.cpython-311-aarch64-linux-gnu.so
#
# venv pkgs => python3 =>
#   import gpiod
#   gpiod.version.__version__ # 2.2.1
#   gpiod # <module 'gpiod' from '/home/pi/wes-gpio/.venv/lib/python3.11/site-packages/gpiod/__init__.py'>
#   help(gpiod)   # completely diff APIs
#
#
# IIUC, as of v2.0.2 they made breaking changes to the APIs that replaced the unofficial pure python APIS (must be the system pkg version b/c it is 1.6.3)... and going forward it sounds like the new APIs are via pypi pkg only (not apt installed cpython pkg)... I bet that is is, the cpython impl is probably not updated to 2.x b/c that would break reality, and so its left at that point in time and won't ever be migrated to 2.x... would make sense
#
#
# I do know the examples in the pypi page (https://pypi.org/project/gpiod/) show using request_lines so def aligns with the fact that I s/b using venv for the following request_lines usages:

import time


def precise_delay_us(us):
    # PRN address clock updates/corrections => only when I detect it as a problem
    # time.sleep is WAY off... hundreds of us in the LA1010 waveform
    start = time.perf_counter()
    while (time.perf_counter() - start) < (us / 1_000_000):
        pass


def wait_for_recovery_between_bits():
    precise_delay_us(3)  # when 2us (sometimes 1us not met b/c took a while to rise back)


def initialize_bus() -> bool:

    # PRN pull high initially and wait Xus to be sure it was high before we pull low?
    with gpiod.request_lines(
            "/dev/gpiochip4",
            consumer="send-init-bus",
            config={DS1820B_PIN: gpiod.LineSettings(direction=Direction.OUTPUT, output_value=LOW)},
    ) as line:
        # initialize pulse is:
        #   480us low => release
        #   wait 15-60us for presence signal from sensor(s)
        #   presence signal (low) lasts 60us-240us

        precise_delay_us(480)  # 480us (max 960us) => MEASURED 545us!? (LA1010)
        line.set_value(DS1820B_PIN, HIGH)

        # poll for presence signal from sensor(s)
        timeout_start_time = time.time()
        while line.get_value(DS1820B_PIN) != LOW:
            if time.time() - timeout_start_time > 0.000_100:  # timeout after 100us b/c should start w/in 15-60us
                logger.error("Presence signal not received")
                return False

        # poll for presence signal to end
        timeout_start_time = time.time()
        while line.get_value(DS1820B_PIN) == LOW:
            if time.time() - timeout_start_time > 0.001:
                logger.error("Presence signal did not end")
                return False

        # wait at least 480us for entirety of presence command + recovery time
        us_since_presence_start = math.floor((time.time() - timeout_start_time) * 1_000_000)
        precise_delay_us(480 - us_since_presence_start)  # wait for 480us total

        return True


def send_command(line, command) -> bool:
    # TODO written for READ ROM command, might need to make changes or other versions for other CMDs

    cmd_bits = []
    # TODO scope the waverform after no comms for a while, seems to initially have failed read rom... and then works after that first round... see if maybe line was not high or?
    # build bits so we can send in left to right order in next loop
    for i in range(8):
        # bits are sent in reverse (confirmed w/ protocol analyzer on LA1010 which successfully matched my READ ROM 0x33 command => )
        this_bit = (command >> i) & 1
        cmd_bits = cmd_bits + [this_bit]  # append each bit to end of list
    for bit in cmd_bits:
        if bit:
            # write 1
            line.set_value(DS1820B_PIN, LOW)
            precise_delay_us(2)  # min 1us, max <15us
            line.set_value(DS1820B_PIN, HIGH)
            precise_delay_us(60)  # 60 us total window (min)
        else:
            # write 0
            line.set_value(DS1820B_PIN, LOW)
            precise_delay_us(65)  # min 60us => wow turned into 120us (LA1010),73us, 68us, 72us ...  120us breaks the rules (max 120)... the rest work inadvertently
            line.set_value(DS1820B_PIN, HIGH)
            # PRN wait for it to be high? I am noticing that when I am low for along time and then go high, it seems to cut into recovery between bits
        wait_for_recovery_between_bits()
    print(f"sent command: {command:08b} ({hex(command)})")
    return True


def read_rom_response(line) -> bool:

    response_bits = []

    def read_bit():
        # line.reconfigure_lines({DS1820B_PIN: gpiod.LineSettings(direction=Direction.OUTPUT, output_value=HIGH)}) # don't need to reconfigure
        line.set_value(DS1820B_PIN, HIGH)  # added after removing reconfigure b/c w/o this reading fails

        # PRN check direction before setting? might only matter on first byte and the overhead here is NBD as nothing timing matters until I pull low
        line.set_value(DS1820B_PIN, LOW)  # host starts the read by driving low for >1us but not long
        start_time = time.time()  # starts after pull low, have up to 15us to read the bit 0/1 for sure though I am seeing 31ish us for 0s, <5us for 1s
        precise_delay_us(1)  # min time 1 us

        # FYI using reconfigure is adding 7-8us of time before 1's can be read so that is bad news here... driving high works fine
        line.set_value(DS1820B_PIN, HIGH)  # fastest response times (~5us for read 1)
        # line.reconfigure_lines({DS1820B_PIN: gpiod.LineSettings(direction=Direction.INPUT)})  # adds (~12+ us for read 1, ouch)

        while line.get_value(DS1820B_PIN) == LOW:
            if time.time() - start_time > 1:
                logger.error("timeout - held low indefinitely - s/b NOT POSSIBLE")
                return False
        end_time = time.time()
        seconds_low = end_time - start_time
        if seconds_low > 0.000_015:
            response_bits.append(0)
        elif seconds_low < 0.000_002:
            # looks like sensor never held it low past me so this is invalid
            logger.error("timeout - sensor not holding low after I release - it is not responding to read request")
            return False
        else:
            response_bits.append(1)
        while time.time() - start_time < 0.000_080:  # TODO did increasing this make reads more reliable? (60us required minimum)
            # all read slots must be 60us (min)
            pass
        wait_for_recovery_between_bits()
        return True

    for i in range(64):
        if (not read_bit()):
            print(f"Failed to read bit {i}, aborting...")
            return False

    print(f"bits read: {response_bits}")
    all_bytes = []
    for i in range(0, 64, 8):
        byte = 0
        for j in range(8):
            byte = byte | (response_bits[i + j] << j)
        all_bytes.append(byte)

    # *** see data sheet:
    #     Then starting with the least significant bit of the family code, 1 bit at a time is shifted in...
    #     are bits in reverse order within each byte?
    #
    #     | 8-BIT CRC CODE |  48-BIT SERIAL NUMBER | 8-BIT FAMILY CODE |
    #     |                |                       |       (28h)       |
    #     | MSB        LSB |  MSB              LSB | MSB           LSB |
    #
    print("bytes:")
    for byte in all_bytes:
        print(f"  {byte:08b} ({byte})")
        # YAY often I am seeing the same bits in each byte... 1st and 5th sometimes vary...

    # Define the CRC-8 function using the polynomial 0x131 (x^8 + x^5 + x^4 + 1)
    crc_all = ds18b20_crc8(bytes(all_bytes))  # if include last byte then it should come out to 0, no need to know CRC computed vs actual if they don't match anyways
    if crc_all != 0:
        print(f"Failed CRC check: {crc_all}")
        return False

    family_code = all_bytes[0]  # 8 bits (1 byte)
    serial_number = all_bytes[1:7]  # 48 bits (6 bytes)
    crc = all_bytes[7]  # 8 bits (1 byte)
    logger.info(f"Serial number: {serial_number}")
    logger.info(f"CRC: {crc}")
    if family_code != 0x28:
        print(f"Invalid family code: {hex(family_code)}, expected 0x28")
        return False

    return True


def wait_for_temp_conversion_to_complete(line):
    timeout_start_time = time.time()
    while line.get_value(DS1820B_PIN) == LOW:
        if time.time() - timeout_start_time > 1:
            logger.error("Temp conversion did not complete after 1sec")
            return False
    return True


ROM_READ_CMD = 0x33
CONVERT_T_CMD = 0x44
READ_SCRATCHPAD_CMD = 0xBE
# THIS IS MY VIDEO today => practical AI use case... let AI lookup values and just confirm them for you...  fill out this list (see page 14/27 for more cmds)


def test_read_temp() -> bool:

    with gpiod.request_lines(
            "/dev/gpiochip4",
            consumer="send-command",
            config={DS1820B_PIN: gpiod.LineSettings(direction=Direction.OUTPUT, output_value=HIGH)},  # FYI CONFIRMED => keep it high for so any overhead in request line isn't adding to total time low on first bit if 0
            #   PREV defaulted to low and that added 50us to the first bit low time!!!!
    ) as line:
        return send_command(line, ROM_READ_CMD)  \
            and read_rom_response(line) \
            and send_command(line, CONVERT_T_CMD) \
            and wait_for_temp_conversion_to_complete(line)


def main():

    if (not initialize_bus()):
        print("Failed to initialize bus")
        return
    print("Bus initialized")
    test_read_temp()


if __name__ == "__main__":
    main()

# NOTES
# ** wait/read_edge_events => my first attempt didn't work?! are these a superior interface?

File: vms/compile-rust/wes-gpio/ds18b20/c/ds18b20.c
#define LOG_LEVEL_INFO 1
#define LOG_LEVEL_DEBUG 2
#define LOG_LEVEL LOG_LEVEL_DEBUG

#include <gpiod.h>
// sudo apt-get install libgpiod-dev
// dpkg -S gpiod.h => /usr/include/gpiod.h
#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include "logging.c"
#include <stdint.h>

#define LOW 0
#define HIGH 1
#define RELEASE HIGH
#define ASSUME_PREV_HIGH HIGH

// ROM commands
#define SKIP_ROM 0xCC
#define READ_ROM 0x33
#define SEARCH_ROM 0xF0
#define MATCH_ROM 0x55
#define ALARM_SEARCH 0xEC

// memory commands (must come after a ROM command)
#define CONVERT_T_CMD 0x44
#define READ_SCRATCHPAD 0xBE
#define RECALL_E2 0xB8
#define WRITE_SCRATCHPAD 0x4E
#define COPY_SCRATCHPAD 0x48
#define MEMORY_READ_POWER_SUPPLY 0xB4 // PRN rename other memory commands that aren't obvious, maybe prefix all with MEMORY_ (that way I can easily spot if I am using commands in the wrong order)

#define GPIO_CHIP_NAME "gpiochip4"
#define GPIO_CHIP_LABEL "pinctrl-rp1"
#define GPIO_LINE_DS18B20 12

bool precise_delay_us(unsigned int us)
{
  struct timespec start, now;
  clock_gettime(CLOCK_MONOTONIC, &start);

  while (1)
  {
    clock_gettime(CLOCK_MONOTONIC, &now);
    long elapsed_us = (now.tv_sec - start.tv_sec) * 1e6 +
                      (now.tv_nsec - start.tv_nsec) / 1e3;
    if (elapsed_us >= us)
    {
      break;
    }
  }
  return true; // just so I can compose with other boolean operations
  // FYI w1 therm driver in kernel impls spin timer => might be more accurate than what I am using above so consider it if more accurate timing is needed
  //  w1_delay https://github.com/raspberrypi/linux/blob/rpi-6.8.y/drivers/w1/w1_io.c#L40-L54
  //    unfortunately it uses udelay (kernel space only, IIUC => linux/asm/delay.h)
  //   https://github.com/raspberrypi/linux/blob/rpi-6.8.y/drivers/w1/w1_io.c#L342-L349
}

bool reset_bus(struct gpiod_line *line)
{
  gpiod_line_set_value(line, LOW);

  precise_delay_us(480);

  gpiod_line_set_value(line, HIGH);

  int start_us = clock();
  while (gpiod_line_get_value(line) == 1)
  {
    if (clock() - start_us > 1000000)
    {
      LOG_ERROR("timeout waiting for line to go low - the start of presence response");
      break;
    }
  }
  // just went low
  start_us = clock();

  while (gpiod_line_get_value(line) == 0)
  {
    if (clock() - start_us > 1000000)
    {
      LOG_ERROR("timeout waiting for line to go high - the end of presence response");
      break;
    }
  }
  // released

  // must wait 480us past start of presence response
  int presence_us = clock() - start_us; // s/b 60 < presence_us < 120
  usleep(480 - presence_us);

  return true;
}

// return the bit that is read (negative if error) (0 is low, 1 is high)
int read_bit(struct gpiod_line *line)
{
  gpiod_line_set_value(line, HIGH); // might be vesitgial

  gpiod_line_set_value(line, LOW); // host starts the read by driving low for >1us but not long
  int start_time = clock();        // starts after pull low, have up to 15us to read the bit 0/1 for sure though I am seeing 31ish us for 0s, <5us for 1s

  // holy cow getting prev_bit (response_bits[-1]) had enough delay to mess up readings so don't use it here
  precise_delay_us(5); // since changing to 3, READ ROM has failed MUCH LESS OFTEN

  gpiod_line_set_value(line, HIGH);

  // simple idea => wait for the line to actually change before reading it (in the event it is a 1 and goes high rather quickly and we have a very very narrow window to read the line during b/c of stupid timing decisions)
  precise_delay_us(2); // give it time to actually change so hopefully don't need to re-read which can push past 15us window to read logic 1s

  while (gpiod_line_get_value(line) == LOW)
  { // ~3us to read
    if (clock() - start_time > 1e6)
    {
      LOG_ERROR("timeout - held low indefinitely - s/b NOT POSSIBLE");
      return -1;
    }
  }

  int us_low = clock() - start_time;

  // ensure 60us on 0 and 15us on 1
  int wait_more_us = 65 - us_low;
  // TODO try 15us on 1s
  if (wait_more_us > 0)
  {
    precise_delay_us(wait_more_us);
  }

  if (us_low > 15)
  {
    return LOW;
  }
  return HIGH;
}

/*  ! TODO either get diff impl of CRC8 or GPL release this code  */
static uint8_t w1_crc8_table[] = {
    0, 94, 188, 226, 97, 63, 221, 131, 194, 156, 126, 32, 163, 253, 31, 65,
    157, 195, 33, 127, 252, 162, 64, 30, 95, 1, 227, 189, 62, 96, 130, 220,
    35, 125, 159, 193, 66, 28, 254, 160, 225, 191, 93, 3, 128, 222, 60, 98,
    190, 224, 2, 92, 223, 129, 99, 61, 124, 34, 192, 158, 29, 67, 161, 255,
    70, 24, 250, 164, 39, 121, 155, 197, 132, 218, 56, 102, 229, 187, 89, 7,
    219, 133, 103, 57, 186, 228, 6, 88, 25, 71, 165, 251, 120, 38, 196, 154,
    101, 59, 217, 135, 4, 90, 184, 230, 167, 249, 27, 69, 198, 152, 122, 36,
    248, 166, 68, 26, 153, 199, 37, 123, 58, 100, 134, 216, 91, 5, 231, 185,
    140, 210, 48, 110, 237, 179, 81, 15, 78, 16, 242, 172, 47, 113, 147, 205,
    17, 79, 173, 243, 112, 46, 204, 146, 211, 141, 111, 49, 178, 236, 14, 80,
    175, 241, 19, 77, 206, 144, 114, 44, 109, 51, 209, 143, 12, 82, 176, 238,
    50, 108, 142, 208, 83, 13, 239, 177, 240, 174, 76, 18, 145, 207, 45, 115,
    202, 148, 118, 40, 171, 245, 23, 73, 8, 86, 180, 234, 105, 55, 213, 139,
    87, 9, 235, 181, 54, 104, 138, 212, 149, 203, 41, 119, 244, 170, 72, 22,
    233, 183, 85, 11, 136, 214, 52, 106, 43, 117, 151, 201, 74, 20, 246, 168,
    116, 42, 200, 150, 21, 75, 169, 247, 182, 232, 10, 84, 215, 137, 107, 53};

uint8_t w1_calc_crc8(uint8_t *data, int len)
{
  uint8_t crc = 0;

  while (len--)
    crc = w1_crc8_table[crc ^ *data++];

  return crc;
}

bool read_bytes(struct gpiod_line *line, uint8_t *bytes, size_t num_bytes)
{
  for (int byteIndex = 0; byteIndex < num_bytes; byteIndex++)
  {
    uint8_t current_byte = 0;
    for (int bitIndex = 0; bitIndex < 8; bitIndex++)
    {
      int bit = read_bit(line);
      if (bit < 0)
      {
        LOG_ERROR("Failed to read byte %d bit %d, aborting...", byteIndex, bitIndex);
        return false;
      }
      // reads: put bit into current_byte at position bitIndex
      // thus we assemble the byte from left to right (reverse of how we read it)
      current_byte = (bit << bitIndex) | current_byte;
    }
    bytes[byteIndex] = current_byte;
  }

  LOG_DEBUG("%d bytes read:", num_bytes);
  for (int i = 0; i < 8; i++)
  {
    // FYI %08b is not standard C, but it is in glibc? IIUC
    LOG_DEBUG("  %08b (0x%02x = %3d)", bytes[i], bytes[i], bytes[i]);
  }

  return true;
}

bool read_bytes_with_crc(struct gpiod_line *line, uint8_t *bytes, size_t num_bytes)
{
  if (!read_bytes(line, bytes, num_bytes))
  {
    return false;
  }

  // FYI crc(bytes) == 0
  //   OR crc(bytes[0:7]) == bytes[8]
  uint8_t crc = w1_calc_crc8(bytes, num_bytes);
  if (crc != 0)
  {
    LOG_ERROR("Failed CRC check: %u", crc);
    return 0; // Or return an appropriate error code
  }
  return true;
}

bool read_rom_response(struct gpiod_line *line)
{
  uint8_t all_bytes[8];
  if (!read_bytes_with_crc(line, all_bytes, 8))
  {
    return false;
  }

  uint8_t family_code = all_bytes[0]; // 8 bits (1 byte)
  uint8_t serial_number[6];           // 48 bits (6 bytes)
  for (int i = 0; i < 6; i++)
  {
    serial_number[5 - i] = all_bytes[i + 1];
  }

  char serial_number_string[13];
  sprintf(serial_number_string, "%02x%02x%02x%02x%02x%02x", serial_number[0], serial_number[1], serial_number[2], serial_number[3], serial_number[4], serial_number[5]);
  LOG_DEBUG("serial number: %s", serial_number_string);

  if (family_code != 0x28)
  {
    LOG_ERROR("Invalid family code %02x, expected 0x28", family_code);
    return false;
  }

  return true;
}

bool read_scratchpad(struct gpiod_line *line)
{
  uint8_t scratchpad[9];
  if (!read_bytes_with_crc(line, scratchpad, 9))
  {
    return false;
  }

  uint8_t temp_lsb = scratchpad[0];
  uint8_t temp_msb = scratchpad[1];
  int16_t temp_raw = (temp_msb << 8) | temp_lsb;
  if (temp_raw & 0x8000)
  {
    temp_raw = -((temp_raw ^ 0xFFFF) + 1);
  }
  float temp_celsius = temp_raw / 16.0;
  LOG_INFO("Temp: %.2f°C", temp_celsius);
  float temp_fahrenheit = temp_celsius * 9 / 5 + 32;
  LOG_INFO("Temp: %.2f°F", temp_fahrenheit);

  // PRN read rest of scratchpad beyond temperature when I need it and build a struct response?

  return true;
}

bool write_bit(struct gpiod_line *line, bool this_bit, bool prev_bit)
{
  gpiod_line_set_value(line, HIGH); // ensure high // TODO do I need this?

  LOG_DEBUG("writing bit: %d", this_bit);
  if (this_bit)
  {
    // write 1
    gpiod_line_set_value(line, LOW);
    // min 1us, max <15us
    if (prev_bit == 0)
    {
      precise_delay_us(2); // 0 => 1 needs less time to pull low again by not as high above threshold
    }
    else
    {
      precise_delay_us(5); // 1 => 1 needs more time to pull low b/c way above threshold
    }
    gpiod_line_set_value(line, HIGH);
    precise_delay_us(60); // 60 us total window (min)
  }
  else
  {
    // write 0
    gpiod_line_set_value(line, LOW);
    precise_delay_us(65); // min 60us => wow turned into 120us (LA1010),73us, 68us, 72us ...  120us breaks the rules (max 120)... the rest work inadvertently
    gpiod_line_set_value(line, HIGH);
    // PRN wait for it to be high? I am noticing that when I am low for along time and then go high, it seems to cut into recovery between bits
  }
  precise_delay_us(3); // recovery >1us
  return true;
}

bool send_command(struct gpiod_line *line, uint8_t command)
{
  // TODO analyze timing of 0=>0, 0=>1, 1=>0, 1=>1 (it looks like each has unique timing that can be perfected probably to avoid issues)
  bool prev_bit = 1; // s/b high and worse case we just wait a smidge longer on first writing first bit when its 1
  for (int i = 0; i < 8; i++)
  {
    // FYI bits are sent in reverse order
    bool this_bit = (command >> i) & 1;
    if (!write_bit(line, this_bit, prev_bit))
    {
      return false; // assume inner write_bit already logged explanation
    }
    prev_bit = this_bit;
  }

  switch (command)
  {

    // ROM commands
  case SKIP_ROM:
    LOG_INFO("sent command: SKIP_ROM 0x%02x", SKIP_ROM);
    break;
  case READ_ROM:
    LOG_INFO("sent command: READ_ROM 0x%02x", READ_ROM);
    break;
  case SEARCH_ROM:
    LOG_INFO("sent command: SEARCH_ROM 0x%02x", SEARCH_ROM);
    break;
  case MATCH_ROM:
    LOG_INFO("sent command: MATCH_ROM 0x%02x", MATCH_ROM);
    break;

    // Memory commands
  case CONVERT_T_CMD:
    LOG_INFO("sent command: CONVERT_T_CMD 0x%02x", CONVERT_T_CMD);
    break;
  case READ_SCRATCHPAD:
    LOG_INFO("sent command: READ_SCRATCHPAD 0x%02x", READ_SCRATCHPAD);
    break;
  case WRITE_SCRATCHPAD:
    LOG_INFO("sent command: WRITE_SCRATCHPAD 0x%02x", WRITE_SCRATCHPAD);
    break;
  case COPY_SCRATCHPAD:
    LOG_INFO("sent command: COPY_SCRATCHPAD 0x%02x", COPY_SCRATCHPAD);
    break;
  case RECALL_E2:
    LOG_INFO("sent command: RECALL_E2 0x%02x", RECALL_E2);
    break;
  case MEMORY_READ_POWER_SUPPLY:
    LOG_INFO("sent command: READ_POWER_SUPPLY 0x%02x", MEMORY_READ_POWER_SUPPLY);
    break;
  default:
    LOG_INFO("sent generic command: %d (0x%02x) - please add command name to list to show meaningful name", command, command);
    break;
  }

  return true;
}

bool wait_for_temp_conversion_to_complete(struct gpiod_line *line)
{
  // FYI if parasite power, need to hold data line HIGH for 750ms (Tconv)
  //   IIAC trying a read slot during parasite power would interrupt and corrupt the Convert T command
  //   if external power, can use read slots to check status
  //   probably best just to always wait 750ms b/c what is 100ms less gonna matter?!
  //

  // FYI if you call Convert T then it writes to memory a temp value and subsequent reads will work without this delay... so the first time you boot up and request temp you need to manually ywait to get around hte 750ms min on 12-bit resolution
  // TODO fix this, why isn't it working? also look at python impl and fix it too if fixed here
  // ! TODO check manual - says smth about 10us to release bus after convert T command... am I doing that? why does this successfully wait for read slot to return 1 after 600ms and then read scratch pad always fails?
  // FYI this issue is not waiting, but READ_SCRATCHPAD always fails

  // WTF, now, waiting for a second results in read scratch pad CRC failures... WTF? without delay I am reading an old value, reliably?!
  // precise_delay_us(1000000);
  return true;
  int start_time = clock();
  while (1)
  {
    int bit = read_bit(line);
    if (bit < 0)
    {
      int elapsed_ms = (clock() - start_time) / 1000;
      LOG_ERROR("Failed to read bit for status of temp conversion, aborting after %d ms", (clock() - start_time) / 1000);
      return false;
    }
    if (bit == 1)
    {
      int elapsed_ms = (clock() - start_time) / 1000;
      LOG_INFO("Temp conversion complete after %d ms", elapsed_ms);
      break;
    }
    if (clock() - start_time > 1e6)
    {
      LOG_ERROR("Timeout waiting for temp conversion to complete");
      return false;
    }
    precise_delay_us(50000); // check every 50ms
  }

  return true;
}

bool search_rom(struct gpiod_line *line)
{
  send_command(line, SEARCH_ROM);

  uint8_t read1 = read_bit(line);
  uint8_t read2 = read_bit(line);
  // | read1 | read2 |
  // |   0   |   0   |   sensor(s) remaining with 0 and 1
  // |   0   |   1   |   sensor(s) remaining with 0, none with 1
  // |   1   |   0   |   sensor(s) remaining with 1, none with 0
  // |   1   |   1   |   no sensors (remaining)

  // create an array of  to store discovered bits and help sequence the search
  return false; // TODO
}

bool read_power_supply_type(struct gpiod_line *line)
{
  bool success = reset_bus(line)                 //
                 && send_command(line, SKIP_ROM) //
                 && send_command(line, MEMORY_READ_POWER_SUPPLY);
  if (!success)
  {
    LOG_ERROR("Failed to send command to read power supply type");
    return false;
  }
  int response = read_bit(line);
  if (response < 0)
  {
    LOG_ERROR("Failed to read power supply status bit");
    return false;
  }
  // PRN extract method that can return power supply type for code branching (i.e. to know to wait 750ms versus reading status bit on Convert T)
  if (response)
  {
    LOG_INFO("External power supply: %d", response);
    return true;
  }
  LOG_INFO("Parasite power (using data line, no external power): %d", response);
  return true;
}

int main()
{

  if (CLOCKS_PER_SEC != 1000000)
  {
    LOG_ERROR("CLOCKS_PER_SEC is not 1_000_000 (us), it is %d", CLOCKS_PER_SEC);
    return 1;
  }

  // struct gpiod_chip *chip = gpiod_chip_open_by_name(GPIO_CHIP_NAME);
  struct gpiod_chip *chip = gpiod_chip_open_by_label(GPIO_CHIP_LABEL);
  if (!chip)
  {
    perror("Open GPIO chip failed");
    return 1;
  }

  struct gpiod_line *line = gpiod_chip_get_line(chip, GPIO_LINE_DS18B20);
  if (!line)
  {
    perror("Get GPIO line failed");
    gpiod_chip_close(chip);
    return 1;
  }

  // FYI also event based interface instead: gpiod_line_request_rising_edge_events, IIUC incompatible with read/write APIs
  int ret = gpiod_line_request_output(line, "my_program", HIGH); // keep high (unchanged initially)
  if (ret < 0)
  {
    perror("Request line as output failed");
    gpiod_chip_close(chip);
    return 1;
  }

  // TODO remove or optimize the precise_delay_us(100) - added to help protocol analyzer identify fields clearly

  // reset_bus(line) && send_command(line, READ_ROM) //
  //     && precise_delay_us(100) && read_rom_response(line);

  // reset_bus(line) && send_command(line, SKIP_ROM)                                 // use comment to stop vcformat from combining lines (hack for now)
  //     && precise_delay_us(100) && send_command(line, CONVERT_T_CMD)               //
  //     && wait_for_temp_conversion_to_complete(line)                               //
  //     && precise_delay_us(100) && reset_bus(line) && send_command(line, SKIP_ROM) //
  //     && precise_delay_us(100) && send_command(line, READ_SCRATCHPAD) && read_scratchpad(line);

  // reset_bus(line) && search_rom(line); // for multi device setups =

  read_power_supply_type(line);

  // cleanup
  gpiod_line_release(line);
  gpiod_chip_close(chip);

  return 0;
}
